### PR TITLE
router: suppress re-render on same state

### LINKED ISSUE #25: Effects shouldn't trigger re-render
According to the [documentation on effects](https://github.com/yoshuawuyts/choo#effects), `Unlike reducers they cannot modify the state`. However, because [send-action](https://github.com/sethvincent/send-action/blob/master/index.js) calls the `onchange` handler every time `send()` is called, choo [re-renders the view](https://github.com/yoshuawuyts/choo/blob/master/index.js#L157-L162), which is of course unnecessary since the state hasn't changed.

[Here is a demo](http://requirebin.com/?gist=a6267d63da432ed2be0eaf2a6ef65f20) that shows the view is re-rendered by the `get` effect. A second later, it gets re-rendered by the `receive` effect, where the state actually changed.

One solution here may be to post an issue in the send-action repo that `onchange` shouldn't be called unless the state's actually changed. But another way to think of it is that choo's effects should not trigger the `onchange` handler at all. Thoughts?

### PR DESCRIPTION
**Pull request by yoshuawuyts**
- views will no longer be re-rendered if the state hasn't changed
- `oldState` is passed as the final argument in a view. E.g. `params,
  state, send, oldState`. This is mostly because it will usually only be
  used for optimizing - changing the order would be a breaking change :(

Closes #25 

Minor version bump

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -132,9 +134,12 @@ function choo () {
       const _reducers = ns ? reducers[ns] : reducers
       if (_reducers && _reducers[action.type]) {
         if (ns) {
-          const newState = _reducers[action.type](action, state[ns])
-          state[ns] = xtend(state[ns], newState)
-        } else state = xtend(state, reducers[action.type](action, state))
+          const redState = _reducers[action.type](action, state[ns])
```
Comment (timwis) [inline index.js]: Took me a minute to figure out what you meant by `redState` - perhaps a name like `reducedState` may improve readability?

**Comment by timwis** [comment]
@yoshuawuyts thanks for jumping on this! I can't quite figure out what the new `mutate` logic is doing differently. My first thought was that, rather than cloning the state and altering it (the immutable way), you were going to operate on the original state object so that you could check to see if it's the same object reference (the `===` operator) but that doesn't appear to be what's happening since you clone `state` into `newState` before-hand.

```javascript
const model = {
  state: {
    value: 0
  },
  reducers: {
    increase: (action, state) => ({value: state += action})
  }
}

send('increase', 5) // should trigger re-render
send('increase', 0) // should not trigger re-render
```
If I'm reading this pull request's changes correctly, both `send`s would trigger re-render, no?

**Comment by yoshuawuyts** [comment]
> If I'm reading this pull request's changes correctly, both sends would trigger re-render, no?

Yup, because a reducer was triggered, in most cases a value in the state would change. But now that the `oldState` is also passed as a tailing argument, it means the path has been opened to optimize further down the tree:
```js
var _view = null
function createView (params, state, send, oldState) {
  if (_view && state.someValue === oldState.someValue) return _view
  _view = choo.view`<div>something ${state.someValue}</div>`
  return _view
}
```
To make it behave consistently lifecycle hooks should probably introduced too so that it can be dumped from memory when it's unmounted; but that's a different issue (#1).

Also not sure about the order of arguments here; also feel like `params` could be appended to state, soooo.... getting to a possible api of `view = (state, oldState, send) => choo.view` is a different concern hah.

**Comment by timwis** [comment]
Interesting - not sure what the changes on L137-142 are for if that's the case, but perhaps those are just unrelated optimizations.

For what it's worth, I would expect it to re-render _only_ if the state _actually_ changed (assuming you can also imagine scenarios where a reducer doesn't actually change the state - say, because the user pressed the "calculate" button without changing the value). I imagine the only way to ensure that is to do a deep comparison of the state object, which I understand is expensive. But I wonder what's more expensive: a deep comparison or the DOM tree comparison that occurs when you run `yo.update()`.

(Hope this is a helpful discussion)

File: index.js (hunk 1 above)
Comment (yoshuawuyts) [inline index.js]: Fair enough

**Comment by yoshuawuyts** [comment]
> But I wonder what's more expensive: a deep comparison or the DOM tree comparison that occurs when you run yo.update().

This is all discussing optimizations for an edge case: the chance that you call a function that modifies state that doesn't end up modifying the state is slim.

The new mutate logic makes sure the old state is persisted as-is, and new changes applied to a fresh object which is then conditionally returned. 

I feel like for now this is the best we can do - gonna merge this :sparkles:

**Comment by timwis** [comment]
Sounds good - sorry to belabor the point. Thanks for addressing this!

**Pull request merged**